main_test.go

```diff
@@ -612,6 +695,20 @@ func TestRotate(t *testing.T) {
 				t.Fatalf("Error reopening spreadsheet: %s", err)
 			}
 
+			// verify number of file uploads to S3
```

Comment:
This is a cool idea, but I think it is maybe not ideal for the purpose. Stepping back a little bit, I think the two things that are important are:
1. The final file is uploaded at the end with all the sheets synchronized.
2. The intermedia file is uploaded with the new password after each rotation.

1 is already verified by the existing test code.

This new code counts uploads, which provides what I would call circumstantial evidence for 2. If the number of uploads is correct then probably it is because the file was uploaded at all the right times. But it is not direct evidence, since there are lots of incorrect behaviors that could result in the same upload count. And it requires a bunch of logic recreating what the code under test does (`expectedSynch`), just to make the counts line up.

I thought of another alternative for verifying 2: we could write code in the `Put` handler that compares the uploaded spreadsheet to the current contents of the file and returns an error (or sets a flag that will fail the test later, whichever you think is better) whenever the password manager sheet shows more than one new or changed password. This would verify 2 more directly, have simpler logic, and would not start failing if the code changes how many uploads it does (e.g. by adding or removing a redundant upload) or, worse, continue passing if the code becomes wrong (e.g. by skipping a necessary upload but adding a redundant one).

Reply:
Good points. I just want to document the details. All comparisons are using the automated sheet in the files: The current file (f.Path) is the source of truth. The uploaded file is stored at fakeClient.LocalPath. Before uploading f.Path, fakeClient.PutObject compares the two files. uploadFile() is called [1] after each successful rotation and also [2] after automated sheet is synch'd to MACFin sheet.  [1] If f.Path contains any usernames with a different password from fakeClient.LocalPath, then there should be only 1. If there is more than 1 username with a password that differs from fakeClient.LocalPath then upload(s) were skipped (an error). [2] If all usernames in f.Path have the same password as in fakeClient.LocalPath, then verify that f.Path has the same number of usernames as fakeClient.LocalPath. If not, then f.Path is not synch'd to fakeClient.LocalPath and an upload was skipped (an error). If f.Path and fakeClient.LocalPath match, then this is a redundant upload (not an error).

Reply:
Just one point. The final file is uploaded only if MACFin passwords are updated. Test case `delete_at_beginning` fails if upload is skipped  after synchronization because rows are deleted and no passwords are rotated.

Reply:
I think I should change the code to upload the final file, regardless of whether passwords were rotated.

Reply:
Not sure I understand this. If we are no longer counting uploads, I don't see why that test case would fail.

Reply:
I made a few commits. [1] remove upload from synchronziation function. [2] always upload file in UpdateMACFinUsers function. [3] add a verify function to fakeClient.PutObject that checks whether an upload was skipped after a rotation. Set a flag if an upload was skipped.

I tested by commenting out the uploadFile() call after a rotation. Tests failed as expected. (And without checking the flag, all tests passed.) Great idea to verify upload on every rotation.